### PR TITLE
Mask tracers for immersed boundaries

### PR DESCRIPTION
**Pull request by whitleyv**
This PR masks tracers inside immersed solids in the same way velocities are already being masked:

`mask_immersed_tracers!(tracers, arch, grid::GFIBG) = Tuple(mask_immersed_field!(c) for c in tracers)`

 Tracers are masked in `update_state!` before filling tracer halos.

This also updates the cylinder validation to run with this immersed boundary method with tracers and analysis. The analysis is mostly for no slip boundary conditions which are not implemented yet (just free slip).

Includes visuals like,

https://user-images.githubusercontent.com/67593861/126657679-1daa523e-c1da-4097-b994-d16966a9b578.mp4

and analysis of the pressure and friction coefficient (known for no slip BC), the normal and tangential velocities, and the tracer conservation, taken as the leakage in the fluid as compared to the initial state. Right now the tracers won't look right. @glwagner has found some problems with the diffusivities and tracers to be fixed

analysis_cylinder_tracer_Nh_350_CenteredSecondOrder

### CODE REVIEW

File: src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl [hunk 1]
```diff
@@ -56,19 +56,19 @@ const GFIBG = ImmersedBoundaryGrid{FT, TX, TY, TZ, G, <:GridFittedBoundary} wher
 #####
 
 # ccc, ffc, fcf
- @inline _viscous_flux_ux(i, j, k, ibg::GFIBG, args...) = conditional_flux_ccc(i, j, k, ibg, viscous_flux_ux, args...)
- @inline _viscous_flux_uy(i, j, k, ibg::GFIBG, args...) = conditional_flux_ffc(i, j, k, ibg, viscous_flux_uy, args...)
- @inline _viscous_flux_uz(i, j, k, ibg::GFIBG, args...) = conditional_flux_fcf(i, j, k, ibg, viscous_flux_uz, args...)
+ #@inline _viscous_flux_ux(i, j, k, ibg::GFIBG, args...) = conditional_flux_ccc(i, j, k, ibg, viscous_flux_ux, args...)
+ #@inline _viscous_flux_uy(i, j, k, ibg::GFIBG, args...) = conditional_flux_ffc(i, j, k, ibg, viscous_flux_uy, args...)
+ #@inline _viscous_flux_uz(i, j, k, ibg::GFIBG, args...) = conditional_flux_fcf(i, j, k, ibg, viscous_flux_uz, args...)
```
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl:61]: We want to keep uncomment these for the sake of this PR, right?

File: src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl [hunk 2]
```diff
@@ -56,19 +56,19 @@ const GFIBG = ImmersedBoundaryGrid{FT, TX, TY, TZ, G, <:GridFittedBoundary} wher
...
  
  # ffc, ccc, cff
- @inline _viscous_flux_vx(i, j, k, ibg::GFIBG, args...) = conditional_flux_ffc(i, j, k, ibg, viscous_flux_vx, args...)
- @inline _viscous_flux_vy(i, j, k, ibg::GFIBG, args...) = conditional_flux_ccc(i, j, k, ibg, viscous_flux_vy, args...)
- @inline _viscous_flux_vz(i, j, k, ibg::GFIBG, args...) = conditional_flux_cff(i, j, k, ibg, viscous_flux_vz, args...)
+ #@inline _viscous_flux_vx(i, j, k, ibg::GFIBG, args...) = conditional_flux_ffc(i, j, k, ibg, viscous_flux_vx, args...)
+ #@inline _viscous_flux_vy(i, j, k, ibg::GFIBG, args...) = conditional_flux_ccc(i, j, k, ibg, viscous_flux_vy, args...)
+ #@inline _viscous_flux_vz(i, j, k, ibg::GFIBG, args...) = conditional_flux_cff(i, j, k, ibg, viscous_flux_vz, args...)
```
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl:66]: Ditto as above.

File: src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl [hunk 3]
```diff
@@ -56,19 +56,19 @@ const GFIBG = ImmersedBoundaryGrid{FT, TX, TY, TZ, G, <:GridFittedBoundary} wher
...
  
  # fcf, cff, ccc
- @inline _viscous_flux_wx(i, j, k, ibg::GFIBG, args...) = conditional_flux_fcf(i, j, k, ibg, viscous_flux_wx, args...)
- @inline _viscous_flux_wy(i, j, k, ibg::GFIBG, args...) = conditional_flux_cff(i, j, k, ibg, viscous_flux_wy, args...)
- @inline _viscous_flux_wz(i, j, k, ibg::GFIBG, args...) = conditional_flux_ccc(i, j, k, ibg, viscous_flux_wz, args...)
+ #@inline _viscous_flux_wx(i, j, k, ibg::GFIBG, args...) = conditional_flux_fcf(i, j, k, ibg, viscous_flux_wx, args...)
+ #@inline _viscous_flux_wy(i, j, k, ibg::GFIBG, args...) = conditional_flux_cff(i, j, k, ibg, viscous_flux_wy, args...)
+ #@inline _viscous_flux_wz(i, j, k, ibg::GFIBG, args...) = conditional_flux_ccc(i, j, k, ibg, viscous_flux_wz, args...)
```
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl:71]: Ditto as above

File: src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl [hunk 4]
```diff
@@ -164,3 +164,5 @@ end
 end
 
 mask_immersed_velocities!(U, arch, grid::GFIBG) = Tuple(mask_immersed_field!(q) for q in U)
+
+mask_immersed_tracers!(tracers, arch, grid::GFIBG) = Tuple(mask_immersed_field!(c) for c in tracers)
```
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl:168]: Huh, I'm not sure why we have `mask_immersed_velocities!`. Note that `mask_immersed_tracers!` is identical to the function for velocities, so maybe we don't need both...

File: src/Models/NonhydrostaticModels/update_nonhydrostatic_model_state.jl [hunk 5]
```diff
@@ -10,6 +11,11 @@ import Oceananigans.TimeSteppers: update_state!
 Update peripheral aspects of the model (halo regions, diffusivities, hydrostatic pressure) to the current model state.
 """
 function update_state!(model::NonhydrostaticModel)
+    
+    # Mask immersed tracers
+    tracer_masking_events = mask_immersed_tracers!(model.tracers, model.architecture, model.grid)
```
Comment (glwagner) [inline src/Models/NonhydrostaticModels/update_nonhydrostatic_model_state.jl:16]:
```suggestion
    tracer_masking_events = Tuple(mask_immersed_field!(c) for c in model.tracers)
```

Then we can get rid of `mask_immersed_tracers!`.

File: src/Models/NonhydrostaticModels/update_nonhydrostatic_model_state.jl [hunk 6]
```diff
@@ -1,6 +1,7 @@
 using Oceananigans.Architectures
 using Oceananigans.BoundaryConditions
 using Oceananigans.TurbulenceClosures: calculate_diffusivities!
+using Oceananigans.ImmersedBoundaries: mask_immersed_tracers!
```
Comment (glwagner) [inline src/Models/NonhydrostaticModels/update_nonhydrostatic_model_state.jl:4]:
```suggestion
using Oceananigans.ImmersedBoundaries: mask_immersed_field!
```

File: validation/immersed_boundaries/Project.toml [hunk 7]
```diff
@@ -1,3 +1,5 @@
 [deps]
+ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
```
Comment (glwagner) [inline validation/immersed_boundaries/Project.toml:3]: Ooh, I don't think we want these. Can we reset `Project.toml` and `Manifest.toml` to master?

**Review by glwagner — COMMENTED** [review]
Couple details to sort out before we merge into the master branch, otherwise the new example looks great!

**Comment by glwagner** [comment]
The plume emerging from the tracer is due to leakage right? It's a nice way to visualize it quickly.

File: src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl (hunk 1 above)
Comment (whitleyv) [reply src/ImmersedBoundaries/grid_fitted_immersed_boundary.jl:61]: ah yes, sorry forgot to change that back!

**Comment by whitleyv** [comment]
> The plume emerging from the tracer is due to leakage right? It's a nice way to visualize it quickly.

Yes. Without the masking, the 0 concentration in the cylinder just leaks out until it's replaced by the fluid concentration. You can see it in this one:

https://user-images.githubusercontent.com/67593861/126666332-df00dec5-3c1c-4d17-9c67-cf7fdc4bd26a.mp4

File: validation/immersed_boundaries/Project.toml (hunk 7 above)
Comment (whitleyv) [reply validation/immersed_boundaries/Project.toml:3]: I thought the idea was to add things like this needed for the validation to the local project inside the validation, and not the main Project.toml?

Comment (glwagner) [reply validation/immersed_boundaries/Project.toml:3]: ah, I missed that this was in `validation/immersed_boundaries/Project.toml`! Apologies.

**Review by glwagner — APPROVED** [review]

**Comment by glwagner** [comment]
Merge away.

**Comment by glwagner** [comment]
> > The plume emerging from the tracer is due to leakage right? It's a nice way to visualize it quickly.
> 
> Yes. Without the masking, the 0 concentration in the cylinder just leaks out until it's replaced by the fluid concentration. You can see it in this one:
> 
>  contours_tracer_Nh_1000_5thOrder.mp4

Hmm, that's another way to visualize the leakage.

**Pull request merged**